File: opendc-simulator/opendc-simulator-compute/src/main/java/org/opendc/simulator/compute/workload/SimTrace.java
/*
 * Copyright (c) 2022 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.simulator.compute.workload;

import java.util.ArrayDeque;
import java.util.Iterator;
import java.util.List;
import org.opendc.simulator.compute.SimMachineContext;
import org.opendc.simulator.compute.SimProcessingUnit;
import org.opendc.simulator.flow2.FlowGraph;
import org.opendc.simulator.flow2.FlowStage;
import org.opendc.simulator.flow2.FlowStageLogic;
import org.opendc.simulator.flow2.OutPort;

/**
 * A workload trace that describes the resource utilization over time in a collection of {@link SimTraceFragment}s.
 */
public final class SimTrace {
    private final ArrayDeque<SimTraceFragment> fragments;
    /**
     * Construct a {@link SimTrace} instance.
     *
     */
    private SimTrace(ArrayDeque<SimTraceFragment> fragments) {
        if (fragments.isEmpty()) {
            throw new IllegalArgumentException("No Fragments found for the Trace");
        }
        this.fragments = fragments;
    }

    /**
     * Construct a {@link SimWorkload} for this trace.
     *
     * //     * @param offset The offset for the timestamps.
     */
    public SimWorkload createWorkload(long start) {
        return createWorkload(start, 0, 0, 1);
    }

    /**
     * Construct a {@link SimWorkload} for this trace.
     *
     * //     * @param offset The offset for the timestamps.
     */
    public SimWorkload createWorkload(
            long start, long checkpointInterval, long checkpointDuration, double checkpointIntervalScaling) {
        return new Workload(start, fragments, checkpointInterval, checkpointDuration, checkpointIntervalScaling);
    }

    //    /**
    //     * Create a new {@link Builder} instance with a default initial capacity.
    //     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Construct a {@link SimTrace} from the specified fragments.
     *
     * @param fragments The array of fragments to construct the trace from.
     */
    public static SimTrace ofFragments(SimTraceFragment... fragments) {
        final Builder builder = builder();

        for (SimTraceFragment fragment : fragments) {
            builder.add(fragment.duration(), fragment.cpuUsage(), fragment.coreCount());
        }

        return builder.build();
    }

    /**
     * Construct a {@link SimTrace} from the specified fragments.
     *
     * @param fragments The fragments to construct the trace from.
     */
    public static SimTrace ofFragments(List<SimTraceFragment> fragments) {
        final Builder builder = builder();

        for (SimTraceFragment fragment : fragments) {
            builder.add(fragment.duration(), fragment.cpuUsage(), fragment.coreCount());
        }

        return builder.build();
    }

    /**
     * Builder class for a {@link SimTrace}.
     */
    public static final class Builder {
        private final ArrayDeque<SimTraceFragment> fragments;

        private boolean isBuilt;

        /**
         * Construct a new {@link Builder} instance.
         */
        private Builder() {
            this.fragments = new ArrayDeque<>();
        }

        /**
         * Add a fragment to the trace.
         *
         * @param duration The timestamp at which the fragment ends (in epoch millis).
         * @param usage The CPU usage at this fragment.
         * @param cores The number of cores used during this fragment.
         */
        public void add(long duration, double usage, int cores) {
            if (isBuilt) {
                recreate();
            }

            fragments.add(new SimTraceFragment(duration, usage, cores));
        }

        /**
         * Build the {@link SimTrace} instance.
         */
        public SimTrace build() {
            isBuilt = true;
            return new SimTrace(fragments);
        }

        /**
         * Clone the columns of the trace.
         *
         * <p>
         * This is necessary when a {@link SimTrace} has been built already, but the user is again adding entries to
         * the builder.
         */
        private void recreate() {
            isBuilt = false;
            this.fragments.clear();
        }
    }

    /**
     * Implementation of {@link SimWorkload} that executes a trace.
     */
    private static class Workload implements SimWorkload {
        private WorkloadStageLogic logic;

        private long offset;

        private final long start;
        private ArrayDeque<SimTraceFragment> fragments;

        private long checkpointInterval; // How long to wait until a new checkpoint is made
        private long checkpointDuration; // How long does it take to make a checkpoint
        private double checkpointIntervalScaling;
        private SimWorkload snapshot;

        private Workload(
                long start,
                ArrayDeque<SimTraceFragment> fragments,
                long checkpointInterval,
                long checkpointDuration,
                double checkpointIntervalScaling) {
            this.start = start;
            this.checkpointInterval = checkpointInterval;
            this.checkpointDuration = checkpointDuration;
            this.checkpointIntervalScaling = checkpointIntervalScaling;

            this.fragments = fragments;

            this.snapshot = this;
        }

        @Override
        public long getCheckpointInterval() {
            return checkpointInterval;
        }

        @Override
        public long getCheckpointDuration() {
            return checkpointDuration;
        }

        @Override
        public double getCheckpointIntervalScaling() {
            return checkpointIntervalScaling;
        }

        @Override
        public void setOffset(long now) {
            this.offset = now;
        }

        @Override
        public void onStart(SimMachineContext ctx) {
            final WorkloadStageLogic logic;
            logic = new SingleWorkloadLogic(ctx, offset, fragments.iterator());
            this.logic = logic;
        }

        @Override
        public void onStop(SimMachineContext ctx) {
            final WorkloadStageLogic logic = this.logic;

            if (logic != null) {
                this.logic = null;
                logic.getStage().close();
            }
        }

        @Override
        public void makeSnapshot(long now) {
            final WorkloadStageLogic logic = this.logic;
            final ArrayDeque<SimTraceFragment> newFragments = this.fragments;

            if (logic != null) {
                int index = logic.getIndex();

                if (index == 0 && (logic.getPassedTime(now) == 0)) {
                    this.snapshot = this;
                    return;
                }

                // Remove all finished fragments
                for (int i = 0; i < index; i++) {
                    newFragments.removeFirst();
                }
            } else {
                return;
            }

            // Reduce the current Fragment to a fragment with the remaining time.
            SimTraceFragment currentFragment = newFragments.pop();
            long passedTime = logic.getPassedTime(now);
            long remainingTime = currentFragment.duration() - passedTime;

            if (remainingTime > 0) {
                SimTraceFragment newFragment =
                        new SimTraceFragment(remainingTime, currentFragment.cpuUsage(), currentFragment.coreCount());

                newFragments.addFirst(newFragment);
            }

            // Add snapshot Fragment
            // TODO: improve CPUUsage and coreCount here
            SimTraceFragment snapshotFragment = new SimTraceFragment(checkpointDuration, 123456, 1);
            newFragments.addFirst(snapshotFragment);

            // Update the logic
            this.logic.updateFragments(newFragments.iterator(), now);

            // remove the snapshot Fragment and update fragments
            newFragments.removeFirst();
            this.fragments = newFragments;

            this.snapshot = new Workload(
                    start, this.fragments, checkpointInterval, checkpointDuration, checkpointIntervalScaling);
        }

        @Override
        public SimWorkload getSnapshot() {
            return this.snapshot;
        }

        @Override
        public void createCheckpointModel() {}
    }

    /**
     * Interface to represent the {@link FlowStage} that simulates the trace workload.
     */
    private interface WorkloadStageLogic extends FlowStageLogic {
        /**
         * Return the {@link FlowStage} belonging to this instance.
         */
        FlowStage getStage();

        long getPassedTime(long now);

        void updateFragments(Iterator<SimTraceFragment> newFragments, long offset);

        /**
         * Return the current index of the workload.
         */
        int getIndex();
    }

    /**
     * Implementation of {@link FlowStageLogic} for just a single CPU resource.
     */
    private static class SingleWorkloadLogic implements WorkloadStageLogic {
        private final FlowStage stage;
        private final OutPort output;
        private int index = 0;

        private final SimMachineContext ctx;

        private Iterator<SimTraceFragment> fragments;
        private SimTraceFragment currentFragment;
        private long startOffFragment;

        private SingleWorkloadLogic(SimMachineContext ctx, long offset, Iterator<SimTraceFragment> fragments) {
            this.ctx = ctx;

            this.fragments = fragments;

            final FlowGraph graph = ctx.getGraph();
            stage = graph.newStage(this);

            final SimProcessingUnit cpu = ctx.getCpu();
            final OutPort output = stage.getOutlet("cpu");
            this.output = output;

            graph.connect(output, cpu.getInput());

            // Start the first Fragment
            this.currentFragment = this.fragments.next();
            this.output.push((float) currentFragment.cpuUsage());
            this.startOffFragment = offset;
        }

        public long getPassedTime(long now) {
            return now - this.startOffFragment;
        }

        @Override
        public void updateFragments(Iterator<SimTraceFragment> newFragments, long offset) {
            this.fragments = newFragments;

            // Start the first Fragment
            this.currentFragment = this.fragments.next();
            this.output.push((float) currentFragment.cpuUsage());
            this.startOffFragment = offset;

            this.index = -1;

            this.stage.invalidate();
        }

        @Override
        public long onUpdate(FlowStage ctx, long now) {
            long passedTime = getPassedTime(now);
            long duration = this.currentFragment.duration();

            // The current Fragment has not yet been finished, continue
            if (passedTime < duration) {
                return now + (duration - passedTime);
            }

            // Loop through fragments until the passed time is filled.
            // We need a while loop to account for skipping of fragments.
            while (passedTime >= duration) {
                if (!this.fragments.hasNext()) {
                    return doStop(ctx);
                }

                passedTime = passedTime - duration;

                // get next Fragment
                this.index++;
                currentFragment = this.fragments.next();
                duration = currentFragment.duration();
            }

            // start new fragment
            this.startOffFragment = now - passedTime;

            // Change the cpu Usage to the new Fragment
            this.output.push((float) currentFragment.cpuUsage());

            // Return the time when the current fragment will complete
            return this.startOffFragment + duration;
        }

        @Override
        public FlowStage getStage() {
            return stage;
        }

        @Override
        public int getIndex() {
            return index;
        }

        /**
         * Helper method to stop the execution of the workload.
         */
        private long doStop(FlowStage ctx) {
            final SimMachineContext machineContext = this.ctx;
            if (machineContext != null) {
                machineContext.shutdown();
            }
            ctx.close();
            return Long.MAX_VALUE;
        }
    }
}

File: opendc-simulator/opendc-simulator-compute/src/main/kotlin/org/opendc/simulator/compute/workload/SimTrace.kt
/*
 * Copyright (c) 2021 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.simulator.compute.workload

import org.opendc.simulator.compute.model.ProcessingUnit
import org.opendc.simulator.flow.FlowConnection
import org.opendc.simulator.flow.FlowSource
import kotlin.math.min

/**
 * A workload trace that describes the resource utilization over time in a collection of [SimTraceFragment]s.
 *
 * @param usageCol The column containing the CPU usage of each fragment (in MHz).
 * @param timestampCol The column containing the starting timestamp for each fragment (in epoch millis).
 * @param deadlineCol The column containing the ending timestamp for each fragment (in epoch millis).
 * @param coresCol The column containing the utilized cores.
 * @param size The number of fragments in the trace.
 */
public class SimTrace(
    private val usageCol: DoubleArray,
    private val timestampCol: LongArray,
    private val deadlineCol: LongArray,
    private val coresCol: IntArray,
    private val size: Int,
) {
    init {
        require(size >= 0) { "Invalid trace size" }
        require(usageCol.size >= size) { "Invalid number of usage entries" }
        require(timestampCol.size >= size) { "Invalid number of timestamp entries" }
        require(deadlineCol.size >= size) { "Invalid number of deadline entries" }
        require(coresCol.size >= size) { "Invalid number of core entries" }
    }

    public companion object {
        /**
         * Construct a [SimTrace] with the specified fragments.
         */
        public fun ofFragments(fragments: List<SimTraceFragment>): SimTrace {
            val size = fragments.size
            val usageCol = DoubleArray(size)
            val timestampCol = LongArray(size)
            val deadlineCol = LongArray(size)
            val coresCol = IntArray(size)

            for (i in fragments.indices) {
                val fragment = fragments[i]
                usageCol[i] = fragment.usage
                timestampCol[i] = fragment.timestamp
                deadlineCol[i] = fragment.timestamp + fragment.duration
                coresCol[i] = fragment.cores
            }

            return SimTrace(usageCol, timestampCol, deadlineCol, coresCol, size)
        }

        /**
         * Construct a [SimTrace] with the specified fragments.
         */
        @JvmStatic
        public fun ofFragments(vararg fragments: SimTraceFragment): SimTrace {
            val size = fragments.size
            val usageCol = DoubleArray(size)
            val timestampCol = LongArray(size)
            val deadlineCol = LongArray(size)
            val coresCol = IntArray(size)

            for (i in fragments.indices) {
                val fragment = fragments[i]
                usageCol[i] = fragment.usage
                timestampCol[i] = fragment.timestamp
                deadlineCol[i] = fragment.timestamp + fragment.duration
                coresCol[i] = fragment.cores
            }

            return SimTrace(usageCol, timestampCol, deadlineCol, coresCol, size)
        }

        /**
         * Create a [SimTrace.Builder] instance.
         */
        @JvmStatic
        public fun builder(): Builder = Builder()
    }

    /**
     * Construct a new [FlowSource] for the specified [cpu].
     *
     * @param cpu The [ProcessingUnit] for which to create the source.
     * @param offset The time offset to use for the trace.
     * @param fillMode The [FillMode] for filling missing data.
     */
    public fun newSource(cpu: ProcessingUnit, offset: Long, fillMode: FillMode = FillMode.None): FlowSource {
        return CpuConsumer(cpu, offset, fillMode, usageCol, timestampCol, deadlineCol, coresCol, size)
    }

    /**
     * An enumeration describing the modes for filling missing data.
     */
    public enum class FillMode {
        /**
         * When a gap in the trace data occurs, the CPU usage will be set to zero.
         */
        None,

        /**
         * When a gap in the trace data occurs, the previous CPU usage will be used.
         */
        Previous
    }

    /**
     * A builder class for a [SimTrace].
     */
    public class Builder internal constructor() {
        /**
         * The columns of the trace.
         */
        private var usageCol: DoubleArray = DoubleArray(16)
        private var timestampCol: LongArray = LongArray(16)
        private var deadlineCol: LongArray = LongArray(16)
        private var coresCol: IntArray = IntArray(16)

        /**
         * The number of entries in the trace.
         */
        private var size = 0

        /**
         * Add the specified [SimTraceFragment] to the trace.
         */
        public fun add(fragment: SimTraceFragment) {
            add(fragment.timestamp, fragment.timestamp + fragment.duration, fragment.usage, fragment.cores)
        }

        /**
         * Add a fragment to the trace.
         *
         * @param timestamp Timestamp at which the fragment starts (in epoch millis).
         * @param deadline Timestamp at which the fragment ends (in epoch millis).
         * @param usage CPU usage of this fragment.
         * @param cores Number of cores used.
         */
        public fun add(timestamp: Long, deadline: Long, usage: Double, cores: Int) {
            val size = size

            if (size == usageCol.size) {
                grow()
            }

            timestampCol[size] = timestamp
            deadlineCol[size] = deadline
            usageCol[size] = usage
            coresCol[size] = cores

            this.size++
        }

        /**
         * Helper function to grow the capacity of the column arrays.
         */
        private fun grow() {
            val arraySize = usageCol.size
            val newSize = arraySize * 2

            usageCol = usageCol.copyOf(newSize)
            timestampCol = timestampCol.copyOf(newSize)
            deadlineCol = deadlineCol.copyOf(newSize)
            coresCol = coresCol.copyOf(newSize)
        }

        /**
         * Construct the immutable [SimTrace].
         */
        public fun build(): SimTrace {
            return SimTrace(usageCol, timestampCol, deadlineCol, coresCol, size)
        }
    }

    /**
     * A CPU consumer for the trace workload.
     */
    private class CpuConsumer(
        cpu: ProcessingUnit,
        private val offset: Long,
        private val fillMode: FillMode,
        private val usageCol: DoubleArray,
        private val timestampCol: LongArray,
        private val deadlineCol: LongArray,
        private val coresCol: IntArray,
        private val size: Int
    ) : FlowSource {
        private val id = cpu.id
        private val coreCount = cpu.node.coreCount

        /**
         * The index in the trace.
         */
        private var _idx = 0

        override fun onPull(conn: FlowConnection, now: Long, delta: Long): Long {
            val size = size
            val nowOffset = now - offset

            var idx = _idx
            val deadlines = deadlineCol
            var deadline = deadlines[idx]

            while (deadline <= nowOffset && ++idx < size) {
                deadline = deadlines[idx]
            }

            if (idx >= size) {
                conn.close()
                return Long.MAX_VALUE
            }

            _idx = idx
            val timestamp = timestampCol[idx]

            // There is a gap in the trace, since the next fragment starts in the future.
            if (timestamp > nowOffset) {
                when (fillMode) {
                    FillMode.None -> conn.push(0.0) // Reset rate to zero
                    FillMode.Previous -> {} // Keep previous rate
                }
                return timestamp - nowOffset
            }

            val cores = min(coreCount, coresCol[idx])
            val usage = usageCol[idx]

            conn.push(if (id < cores) usage / cores else 0.0)
            return deadline - nowOffset
        }
    }
}
